File: examples/rdma/python/py_rdma/s3rdmaget.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""s3rdmaget.py - download an S3 object via RDMA, using rdmalib + s3client.
Python 2.7 / 3.x compatible.

Usage:
  python s3rdmaget.py -bucket <bucket> -object <key> -file <path>
                       [-endpoint http://127.0.0.1:8901] [-rdma-dev rxe0]
                       [-ak <access-key>] [-sk <secret-key>]
"""
from __future__ import print_function

import argparse
import os
import sys

try:                       # Python 3
    from urllib.parse import urlsplit
except ImportError:        # Python 2
    from urlparse import urlsplit

import rdmalib
import s3client


def ws_url_from_endpoint(endpoint):
    parts = urlsplit(endpoint)
    return "ws://%s/rdma-ctrl" % (parts.netloc,)


def main():
    p = argparse.ArgumentParser()
    p.add_argument("-bucket", required=True, help="Name of the bucket")
    p.add_argument("-object", required=True, help="Name of the object")
    p.add_argument("-file", required=True, help="Path to save the downloaded file")
    p.add_argument("-endpoint", default="http://127.0.0.1:8901", help="S3 endpoint URL")
    p.add_argument("-rdma-dev", dest="rdma_dev", default="rxe0", help="RDMA device name")
    p.add_argument("-ak", default="admin-ak", help="S3 access key")
    p.add_argument("-sk", default="admin-sk", help="S3 secret key")
    args = p.parse_args()

    cfg = s3client.S3Config(endpoint=args.endpoint, region="us-east-1",
                             access_key=args.ak, secret_key=args.sk)

    # Step 1: HEAD to get object size
    print("Getting object info first...", file=sys.stderr)
    try:
        file_size = s3client.head_object(cfg, args.bucket, args.object)
    except Exception as e:
        print("Failed to get object info: %s" % e, file=sys.stderr)
        return 1
    print("Object size: %d bytes" % file_size, file=sys.stderr)

    request_id = "rdma-%d" % os.getpid()

    # Step 2 & 3: RDMA handshake (register, QP setup, mmap buffer, send token)
    ws_url = ws_url_from_endpoint(args.endpoint)
    print("Initializing RDMA resources and exchanging QP info...", file=sys.stderr)
    try:
        sess = rdmalib.RDMASession.recv_setup(ws_url, args.rdma_dev, rdmalib.RDMA_PORT,
                                               rdmalib.RDMA_GID_IDX, request_id, file_size)
    except Exception as e:
        print("RDMA setup failed: %s" % e, file=sys.stderr)
        return 1

    # Step 4: trigger the server-side RDMA write via S3 GetObject
    print("Sending S3 GetObject request...", file=sys.stderr)
    try:
        s3client.get_object(cfg, args.bucket, args.object,
                             ("X-RDMA-Request-ID", request_id))
    except Exception as e:
        print("GetObject error: %s" % e, file=sys.stderr)
        sess.destroy()
        return 1

    print("Successfully requested %s from s3://%s/%s via RDMA (HTTP 200 OK)" %
          (args.file, args.bucket, args.object), file=sys.stderr)

    # Step 5: RDMA write is complete now; copy the data out before releasing resources
    data = sess.buf[:file_size]
    if not isinstance(data, bytes):
        data = bytes(data)
    sess.destroy()

    # Step 6: verify CRC and save to file
    crc = rdmalib.crc32(data)
    print("Data CRC32: 0x%08x" % crc, file=sys.stderr)

    print("Saving data to %s..." % args.file, file=sys.stderr)
    with open(args.file, "wb") as f:
        f.write(data)

    print("Download complete! File saved to %s" % args.file, file=sys.stderr)
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: examples/rdma/java/src/main/java/com/example/s3rdma/S3RdmaGet.java
package com.example.s3rdma;

import java.net.URI;
import java.nio.ByteBuffer;
import java.nio.file.Files;
import java.nio.file.Paths;

/**
 * s3rdmaget - download an S3 object via RDMA, using RdmaSession + S3Client.
 *
 * Usage:
 *   java -Djna.library.path=native -cp target/s3rdma.jar com.example.s3rdma.S3RdmaGet \
 *        -bucket <bucket> -object <key> -file <path>
 *        [-endpoint http://127.0.0.1:8901] [-rdma-dev rxe0] [-ak <ak>] [-sk <sk>]
 */
public class S3RdmaGet {

    private static void usage() {
        System.err.println("Usage: S3RdmaGet -bucket <bucket> -object <key> -file <path> "
                + "[-endpoint http://127.0.0.1:8901] [-rdma-dev rxe0] [-ak <ak>] [-sk <sk>]");
        System.exit(1);
    }

    private static String wsUrlFromEndpoint(String endpoint) throws Exception {
        URI u = URI.create(endpoint);
        String authority = u.getAuthority();
        return "ws://" + authority + "/rdma-ctrl";
    }

    public static void main(String[] args) throws Exception {
        String bucket = null, object = null, filePath = null;
        String endpoint = "http://127.0.0.1:8901";
        String rdmaDev = "rxe0";
        String ak = "admin-ak", sk = "admin-sk";

        for (int i = 0; i < args.length; i++) {
            switch (args[i]) {
                case "-bucket": bucket = args[++i]; break;
                case "-object": object = args[++i]; break;
                case "-file": filePath = args[++i]; break;
                case "-endpoint": endpoint = args[++i]; break;
                case "-rdma-dev": rdmaDev = args[++i]; break;
                case "-ak": ak = args[++i]; break;
                case "-sk": sk = args[++i]; break;
                default: usage();
            }
        }
        if (bucket == null || object == null || filePath == null) {
            usage();
        }

        S3Client cfg = new S3Client(endpoint, "us-east-1", ak, sk);

        // Step 1: HEAD to get object size
        System.err.println("Getting object info first...");
        long fileSize;
        try {
            fileSize = cfg.headObject(bucket, object);
        } catch (Exception e) {
            System.err.println("Failed to get object info: " + e.getMessage());
            System.exit(1);
            return;
        }
        System.err.println("Object size: " + fileSize + " bytes");

        String requestId = "rdma-" + ProcessHandle.current().pid();

        // Step 2 & 3: RDMA handshake (register, QP setup, register MR, send token)
        String wsUrl = wsUrlFromEndpoint(endpoint);
        System.err.println("Initializing RDMA resources and exchanging QP info...");
        RdmaSession sess;
        try {
            sess = RdmaSession.recvSetup(wsUrl, rdmaDev, RdmaSession.RDMA_PORT, RdmaSession.RDMA_GID_IDX,
                    requestId, fileSize);
        } catch (Exception e) {
            System.err.println("RDMA setup failed: " + e.getMessage());
            System.exit(1);
            return;
        }

        // Step 4: trigger the server-side RDMA write via S3 GetObject
        System.err.println("Sending S3 GetObject request...");
        try {
            cfg.getObject(bucket, object, "X-RDMA-Request-ID", requestId);
        } catch (Exception e) {
            System.err.println("GetObject error: " + e.getMessage());
            sess.destroy();
            System.exit(1);
            return;
        }

        System.err.println("Successfully requested " + filePath + " from s3://" + bucket + "/" + object
                + " via RDMA (HTTP 200 OK)");

        // Step 5: RDMA write is complete now; copy the data out before releasing resources
        byte[] data = new byte[(int) fileSize];
        ByteBuffer buf = sess.getBuffer();
        buf.rewind();
        buf.get(data);
        sess.destroy();

        // Step 6: verify CRC and save to file
        long crc = RdmaSession.crc32(data);
        System.err.println(String.format("Data CRC32: 0x%08x", crc));

        System.err.println("Saving data to " + filePath + "...");
        Files.write(Paths.get(filePath), data);

        System.err.println("Download complete! File saved to " + filePath);
    }
}
